addons/purchase/models/product.py

```diff
@@ -78,6 +78,41 @@ def action_view_po(self):
         }
         return action
 
+    @api.model
+    def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_uid=None):
+        if name and self.env.context.get('purchase_favorites'):
+            partner_id = self.env.context.get('partner_id')
+            if partner_id:
+                if not args:
+                    args = []
+                company_id = self.env.context.get('company_id')
+                date_from = fields.Date.context_today(self) - relativedelta(years=1)
+                done_states = ['purchase', 'done']
+                domain = [
+                    ('partner_id', '=', partner_id),
+                    ('date_order', '>=', date_from),
+                    ('state', 'in', done_states),
+                    ('company_id', '=', company_id),
+                ]
+                read_group_result = self.env['purchase.report'].read_group(domain, ['qty_ordered'], 'product_id', limit=None)
+                purchased_product_ids = [purchase_report_line['product_id'][0] for purchase_report_line in read_group_result]
+                purchased_quantities_by_product = {purchase_report_line['product_id'][0]: purchase_report_line['qty_ordered'] for purchase_report_line in read_group_result}
+                in_args = args.copy()
+                in_args.append(('id', 'in', purchased_product_ids))
+                name_search_result = super(ProductProduct, self)._name_search(name, in_args, operator, None, name_get_uid)
+                weighted_name_search_result = [[purchased_quantities_by_product[product_id], product_id] for product_id in name_search_result]
```

Comment:
`purchased_quantities_by_product[product_id]` => `purchased_quantities_by_product.get(product_id, 0.0)` to avoid traceback if product never purchase is found.

Reply:
seems it can't occur : name_search_result has in_args with 'id in purchased_product_ids' (coming from read_group_result also used to fill purchased_quantities_by_product)
Work in 2 steps
first: get matching_purchased_product_ids ordered by purchases/sales, favorites & alphabetical
second(if needed): fill with matching_not_purchased_product_ids ('id not in ...') ordered by favorites & alphabetical

Reply:
It can occur because the `_name_search` in product/model/product.py is a mess and the args is not used in the 
last case `if not product_ids and self._context.get('partner_id'):`.

Reply:
Moreover, the `_name_search` in product/model/product.py seems broken, something it return a Query object instead of ids... We should fix it.